### ISSUE TITLE
Problem with different hash algos

### ISSUE THREAD

**Issue by ArneBachmann** [pypiserver/pypiserver#265]
I uploaded my first package to a self-hosted pypiserver.
Then I tried to install it via pip install, and it failed with
```
ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE
        Expected sha256 1ef075cf79165a57fc0a8c4e11645fcc
             Got        1ef075cf79165a57fc0a8c4e11645fcc585430b8ec5ae5f390308dd5dfcaaacf
```
which makes no sense, as I don't use requirements files.

I debugged pip for an hour and finally found out where the expected hash is coming from:

It stems from the webserver-output in `_app.py` that renders the `pypserver/simple/mypackage` list of files.

I then looked into `pypiserver` source and found the bug:

In `core.py` I have to remove `[:32]` from `_digest_file` and replace `'%s#%s=%.32s'` with `'%s#%s=%s'`.

After that, I can use SHA hashes of any size (including the SHA256 default).

**Comment by mplanchard** [comment]
Thanks for the report! I’ll look into this.

It seems odd to me that pip only wants the first 32 chars of the hash. The
simple repository interface PEP (503) makes no reference to truncating the
hashes:
https://www.python.org/dev/peps/pep-0503/

Could you provide some more detail when you have a moment regarding how
you’re generating and uploading the package, as well as which version of
pip you’re using?

**Comment by mplanchard** [comment]
Also out of curiosity, have you tried this no-cache solution?
https://stackoverflow.com/questions/40183108/python-packages-hash-not-matching-whilst-installing-using-pip

**Comment by ArneBachmann** [comment]
Yes, my first thought was something off with my pip installation.
But it turns out the problem is in pypiserver as detailed above. I can send a pull request tomorrow, if you like.

**Comment by mplanchard** [comment]
Yep, you're absolutely right. I was able to reproduce this all the way back to pip 8.0, and I looked back in the git history, and it appears to have been broken for at least four years! Great find.

I've got a breaking test here locally, so I'll just clean that up and get things updated.

**Comment by ArneBachmann** [comment]
I was questioning my sanity this afternoon, but good to hear my findings were not completely off.
FYI we are using `pypiserver` in a Vagrant VM, if you're interested in that.

**Comment by mplanchard** [comment]
Sorry about that! I was skeptical because the related code hadn't changed in so long, but I guess it's been broken for a while. I just merged the fix into master. Would you mind giving it a shot and verifying it fixes your issue?

**Comment by mplanchard** [comment]
Closing, as this was resolved by #266 and released this evening. Please feel free to open a new issue if you find any problems, and thanks again for the detailed report!
